src/FwdState.cc

```diff
@@ -470,7 +470,7 @@ FwdState::fail(ErrorState * errorState)
     if (pconnRace == racePossible) {
         debugs(17, 5, HERE << "pconn race happened");
         pconnRace = raceHappened;
-        destinations->retryPath(serverConn);
+        destinations->retryPrimePath(serverConn);
```

Comment:
Technically, `serverConn` may be a spare address (from Happy Eyeballs point of view) that was connected first (before the prime address got connected or after the prime address failed to connect). I am concerned that by assuming that this is a prime address, we may reorder destination addresses for the same peer (if the peer still has addresses in both families left).

Reply:
I agree that the retrying address may be 'spare' here but I doubt we can distinguish 'spare' from 'prime' at this point because this classification is wholly HappyConnOpener's business. We could not rely on grouping addresses of the same families in 'destinations' (i.e., place 'serverConn' into 'destinations' near other addresses of the same family). For example, if 'destinations' contains only IPv4 addresses (for the given destination) and we are retrying IPv6 address it is unclear whether to place it before or after IPv4 addresses (i.e., whether IPv6 was 'prime' or 'spare' when it was recently fetched to FwdState). It seems that the only way to overcome this difficulty is to store prime/spare attribute somewhere within the Connection itself.
Note that this problem is old and unrelated to the PR changes.

Reply:
Given the small amount/low complexity of the current PR changes, the likely significant overlap in final solutions, and the related nature of both bugs (i.e. discarding viable destinations and reordering viable destinations), I think we should fix both bugs in this PR.

> It seems that the only way to overcome this difficulty is to store prime/spare attribute somewhere within the Connection itself.

I am sure there are better solutions than polluting innocent and already overburdened Connection. I am also worried about preserving the "spare or prime" classification beyond a single HappyConnOpener instance lifetime -- the very last address for a peer ought to become "prime" if it reenters Happy Eyeballs on a retry, even if it was originally extracted as "spare".

----

How about _never removing_ candidates from the  ResolvedPeers::paths_  _and_ storing a boolean "available" marker (together with Connection) in ConnectionList::value_type?

The new ConnectionList::value_type can be called ResolvedPeerPath, for example (I do not recommend ResolvedPeer because that would falsely imply that we are storing a CachePeer pointer).

The existing ResolvedPeers methods will never return unavailable items. ResolvedPeers::extractFound() will set the new marker to false. The ResolvedPeers::retry*() methods will reset the marker back to true after searching for a matching Connection (using a raw pointer match -- the Connection object itself could not have changed because it is refcounted and never created in the contexts we are interested in, right?).

If needed, we can optimize candidate extraction searches by caching the offset of the last available candidate seen. Without that optimization, one could argue that the lack of "shift" in this solution may create a potentially long linear search. The counter argument is that very similar searches are already possible (e.g., when a spare address is behind a very long list of not-yet-tried prime addresses). I probably would not rush with this error-prone optimization unless we think it is needed.

Reply:
Your idea looks good. Also, if we never remove elements and only append them into ResolvedPeers::paths_ , we can use std::vector instead of std::list as a container type.

Reply:
Sorry, the container type is 'vector' already. I misinterpreted 'ConnectionList' type name...

Reply:
> How about never removing candidates from the ResolvedPeers::paths_ and storing a boolean "available" marker 

Done (03f0268-09e4d7e).

Reply:
I found bugs in the implementation and fixed them (up to 0d359e5).